upup/pkg/fi/cloudup/hetznertasks/server.go

```diff
@@ -46,123 +51,131 @@ type Server struct {
 	Labels map[string]string
 }
 
-var _ fi.CompareWithID = &Server{}
-
-func (v *Server) CompareWithID() *string {
-	return fi.String(strconv.Itoa(fi.IntValue(v.ID)))
-}
-
-func (v *Server) Find(c *fi.Context) (*Server, error) {
+func (v *ServerGroup) Find(c *fi.Context) (*ServerGroup, error) {
 	cloud := c.Cloud.(hetzner.HetznerCloud)
 	client := cloud.ServerClient()
 
-	// TODO(hakman): Find using label selector
-	servers, err := client.All(context.TODO())
+	labelSelector := []string{
+		fmt.Sprintf("%s=%s", hetzner.TagKubernetesClusterName, c.Cluster.Name),
+		fmt.Sprintf("%s=%s", hetzner.TagKubernetesInstanceGroup, v.Labels[hetzner.TagKubernetesInstanceGroup]),
+	}
+	listOptions := hcloud.ListOpts{
+		PerPage:       50,
+		LabelSelector: strings.Join(labelSelector, ","),
+	}
+	serverListOptions := hcloud.ServerListOpts{ListOpts: listOptions}
+	servers, err := client.AllWithOpts(context.TODO(), serverListOptions)
```

Comment:
Unrelated: We should probably just put the context.Context into *fi.Context; there's some guidance that we shouldn't do so but I think it's fine, particularly if the wrapper object doesn't outlast the context.Context.